src/Compiler/AbstractIL/ilread.fs

```diff
@@ -2112,9 +2112,64 @@ and typeDefReader ctxtH : ILTypeDefStored =
         let layout = typeLayoutOfFlags ctxt mdv flags idx
 
         let hasLayout =
-            (match layout with
-             | ILTypeDefLayout.Explicit _ -> true
-             | _ -> false)
+            match layout with
+            | ILTypeDefLayout.Explicit _ -> true
+            | _ -> false
+
+        let containsExtensionMethods =
+            let mutable containsExtensionMethods = false
+            let searchedKey = TaggedIndex(hca_TypeDef, idx)
+
+            let attributesSearcher =
+                { new ISeekReadIndexedRowReader<CustomAttributeRow, CustomAttributeRow, CustomAttributeRow> with
+                    member _.GetRow(i, row) =
+                        seekReadCustomAttributeRow ctxt mdv i &row
+
+                    member _.GetKey(row) = row
+                    member _.CompareKey(key) = hcaCompare searchedKey key.parentIndex
+                    member _.ConvertRow(row) = row
+                }
+
+            let attrsStartIdx, attrsEndIdx =
+                seekReadIndexedRowsRange
+                    (ctxt.getNumRows TableNames.CustomAttribute)
+                    (isSorted ctxt TableNames.CustomAttribute)
+                    attributesSearcher
+
+            let hasAttributes = attrsStartIdx > 0 && attrsEndIdx >= attrsStartIdx
+            let mutable attrIdx = attrsStartIdx
+
+            while hasAttributes && attrIdx <= attrsEndIdx && not containsExtensionMethods do
+                let mutable attr = Unchecked.defaultof<_>
+                attributesSearcher.GetRow(attrIdx, &attr)
+                let attrCtorIdx = attr.typeIndex.index
+
+                let name =
+                    if attr.typeIndex.tag = cat_MethodDef then
+                        let idx = seekMethodDefParent ctxt attrCtorIdx
+                        let _, nameIdx, namespaceIdx, _, _, _ = seekReadTypeDefRow ctxt idx
+                        readBlobHeapAsTypeName ctxt (nameIdx, namespaceIdx)
+                    else
+                        let mrpTag, _, _ = seekReadMemberRefRow ctxt mdv attrCtorIdx
+
+                        if mrpTag.tag <> mrp_TypeRef then
+                            ""
+                        else
+                            let _, nameIdx, namespaceIdx = seekReadTypeRefRow ctxt mdv mrpTag.index
+                            readBlobHeapAsTypeName ctxt (nameIdx, namespaceIdx)
+
+                if name = "System.Runtime.CompilerServices.ExtensionAttribute" then
```

Comment:
We should have a better mechanism of reading attributes (via TcGlobals) probably. Does it not work in this case?

Reply:
> Probably not, since we just reading it from the blob?

Exactly. Here we want a simple way to scan attributes on a type for the presence of something similar to `ExtensionAttribute`, so that we can later request a complete reading of the attributes with the allocation of all necessary `ILAttribute` objects and checking them via `TyconRefHasAttribute g m g.attrib_ExtensionAttribute` only for corresponding types.